File: app/src/main/java/com/productivity/habits/ui/daily/DirectNumericInputDialog.kt
package com.productivity.habits.ui.daily

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp

@Composable
fun DirectNumericInputDialog(
    habitTitle: String,
    currentValue: Double,
    targetValue: Double,
    unit: String?,
    onDismiss: () -> Unit,
    onConfirm: (Double) -> Unit
) {
    var textValue by remember {
        val initialText = if (currentValue % 1.0 == 0.0) {
            currentValue.toLong().toString()
        } else {
            currentValue.toString()
        }
        mutableStateOf(initialText)
    }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                text = "Log $habitTitle",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold
            )
        },
        text = {
            Column(modifier = Modifier.fillMaxWidth()) {
                val targetText = if (targetValue % 1.0 == 0.0) targetValue.toLong().toString() else targetValue.toString()
                Text(
                    text = "Target: $targetText ${unit ?: ""}".trim(),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(12.dp))
                OutlinedTextField(
                    value = textValue,
                    onValueChange = { input ->
                        if (input.isEmpty() || input.matches(Regex("""^\d*\.?\d*$"""))) {
                            textValue = input
                        }
                    },
                    label = { Text(unit?.let { "Value ($it)" } ?: "Value") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(
                        keyboardType = KeyboardType.Decimal,
                        imeAction = ImeAction.Done
                    ),
                    keyboardActions = KeyboardActions(
                        onDone = {
                            val parsed = textValue.toDoubleOrNull() ?: 0.0
                            onConfirm(parsed)
                        }
                    ),
                    modifier = Modifier.fillMaxWidth()
                )
            }
        },
        confirmButton = {
            TextButton(
                onClick = {
                    val parsed = textValue.toDoubleOrNull() ?: 0.0
                    onConfirm(parsed)
                }
            ) {
                Text("Save", fontWeight = FontWeight.Bold)
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("Cancel")
            }
        }
    )
}

File: lib/ui/daily/dialogs/direct_numeric_input_dialog.dart
import 'package:flutter/material.dart';

class DirectNumericInputDialog extends StatefulWidget {
  final String habitTitle;
  final double currentValue;
  final double targetValue;
  final String? unit;
  final VoidCallback onDismiss;
  final ValueChanged<double> onConfirm;

  const DirectNumericInputDialog({
    super.key,
    required this.habitTitle,
    required this.currentValue,
    required this.targetValue,
    this.unit,
    required this.onDismiss,
    required this.onConfirm,
  });

  @override
  State<DirectNumericInputDialog> createState() =>
      _DirectNumericInputDialogState();
}

class _DirectNumericInputDialogState extends State<DirectNumericInputDialog> {
  late final TextEditingController _controller;

  @override
  void initState() {
    super.initState();
    final initialText = widget.currentValue % 1.0 == 0.0
        ? widget.currentValue.toInt().toString()
        : widget.currentValue.toString();
    _controller = TextEditingController(text: initialText);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _submit() {
    final parsed = double.tryParse(_controller.text.trim()) ?? 0.0;
    widget.onConfirm(parsed);
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final targetText = widget.targetValue % 1.0 == 0.0
        ? widget.targetValue.toInt().toString()
        : widget.targetValue.toString();
    final unitText = widget.unit != null && widget.unit!.trim().isNotEmpty
        ? ' ${widget.unit!.trim()}'
        : '';

    return AlertDialog(
      title: Text(
        'Log ${widget.habitTitle}',
        style: theme.textTheme.titleMedium?.copyWith(
          fontWeight: FontWeight.w600,
        ),
      ),
      content: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Target: $targetText$unitText',
            style: theme.textTheme.bodySmall?.copyWith(
              color: theme.colorScheme.onSurfaceVariant,
            ),
          ),
          const SizedBox(height: 12),
          TextField(
            controller: _controller,
            autofocus: true,
            keyboardType:
                const TextInputType.numberWithOptions(decimal: true),
            textInputAction: TextInputAction.done,
            onSubmitted: (_) => _submit(),
            decoration: InputDecoration(
              labelText: widget.unit != null && widget.unit!.isNotEmpty
                  ? 'Value (${widget.unit})'
                  : 'Value',
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(12),
              ),
            ),
          ),
        ],
      ),
      actions: [
        TextButton(
          onPressed: widget.onDismiss,
          child: const Text('Cancel'),
        ),
        TextButton(
          onPressed: _submit,
          child: const Text(
            'Save',
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
        ),
      ],
    );
  }
}
